### PR TITLE
Improve Npgsql health check design

### PR DESCRIPTION
**Pull request by adamsitnik**
This PR implements #2113 for Npgsql:

- offer two public `NpgSqlHealthCheckOptions` ctors: one that requires `ConnectionString` and another that requires `DataSource`. This allows us to ensure that every instance of this type is valid.
- make `ConnectionString` and `DataSource` properties readonly. This allows us to ensure that the customers are not providing both.
- add README.md and explain what we recommend and why

### CODE REVIEW

File: src/HealthChecks.NpgSql/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+## PostgreSQL Health Check
+
+This health check verifies the ability to communicate with [PostgreSQL](https://www.postgresql.org/). It uses the [Npgsql](https://www.npgsql.org/) library.
+
+## NpgsqlDataSource
```
Comment (adamsitnik) [inline src/HealthChecks.NpgSql/README.md:5]: @roji @NinoFloris PTAL at what I wrote here and let me know if this is correct

**Comment by adamsitnik** [comment]
I was unable to get a review, so I've eyeballed it multiple times. It LGTM, so I am merging it. In case I am wrong, we are going to release a fix later.

File: src/HealthChecks.NpgSql/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+Starting with Npgsql 7.0 (and .NET 7), the starting point for any database operation is [NpgsqlDataSource](https://www.npgsql.org/doc/api/Npgsql.NpgsqlDataSource.html). The data source represents your PostgreSQL database, and can hand out connections to it, or support direct execution of SQL against it. The data source encapsulates the various Npgsql configuration needed to connect to PostgreSQL, as well as the **connection pooling which makes Npgsql efficient**.
```
Comment (roji) [inline src/HealthChecks.NpgSql/README.md:7]: I'd maybe go softer and say "the recommended starting point", as connection strings are still fully supported etc.

File: src/HealthChecks.NpgSql/README.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+```csharp
+void Configure(IServiceCollection services)
+{
+    services.AddNpgsqlDataSource("Host=pg_server;Username=test;Password=test;Database=test");
+    services.AddHealthChecks().AddNpgSql();
+}
+```
+
+By default, the `NpgsqlDataSource` instance is resolved from service provider. If you need to access more than one PostgreSQL database, you can use keyed DI services to achieve that:
```
Comment (roji) [inline src/HealthChecks.NpgSql/README.md:27]: :+1:

File: src/HealthChecks.NpgSql/README.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+We encourage you to use [Npgsql.DependencyInjection](https://www.nuget.org/packages/Npgsql.DependencyInjection) package for registering a singleton factory for `NpgsqlDataSource`.  It allows easy configuration of your Npgsql connections and registers the appropriate services in your DI container.
+
+To make the shift to `NpgsqlDataSource` as easy as possible,  the `Npgsql.DependencyInjection` package registers not just a factory for the data source, but also factory for `NpgsqlConnection`  (and even `DbConnection`). So, your app does not need to suddenly start using `NpgsqlDataSource` everywhere.
+
+```csharp
+void Configure(IServiceCollection services)
+{
+    services.AddNpgsqlDataSource("Host=pg_server;Username=test;Password=test;Database=test");
+    services.AddHealthChecks().AddNpgSql();
```
Comment (roji) [inline src/HealthChecks.NpgSql/README.md:23]: Yep, that's the ideal way things should work - the data source is registered independently as a singleton in DI (via Npgsql.DependencyInjection), and then is implicitly picked up by anyone needing it (health checks, EF Core...). A bit similar to an ILoggerFactory.

**Review by roji — COMMENTED** [review]
Sorry for the super late review, but overall everything here seems good! Yeah, the important thing is to have health checks automatically resolve a singleton NpgsqlDataSource from DI (previously registered via Npgsql.DependencyInjection), and for other cases, to provide ways to register with either a provided data source or a connection string.

**Pull request merged**